// @carbon/charts/test-tools.js
var _this = this;
// Extensions
Object.prototype["containsKeysAndValuesFrom"] = function (sourceObject) {
    var result = true;
    Object.keys(sourceObject).map(function (key) {
        var supposedValue = sourceObject[key];
        if (_this[key] !== supposedValue) {
            result = false;
        }
    });
    return true;
};
// Functions
export var createClassyContainer = function (chartType) {
    var classyContainer = document.createElement("div");
    classyContainer.id = "classy-" + chartType + "-chart-holder";
    classyContainer.classList.add("chart-holder");
    return classyContainer;
};
export var grabClassyContainer = function (chartType) {
    return document.getElementById("classy-" + chartType + "-chart-holder");
};
export var removeChart = function (chartType) {
    var oldClassyContainer = grabClassyContainer(chartType);
    if (oldClassyContainer) {
        oldClassyContainer.parentNode.removeChild(oldClassyContainer);
    }
};
export var inputAndProcessedDataMatch = function (chartObj, inputData) {
    // Gather all pieces of processed data for every type of chart
    // const processedDataPoints = chartObj.displayData.reduce((finalArray, dataPoint) => {
    // 	if (dataPoint.items) {
    // 		// this is the "Other" label, in Pie & Donut
    // 		dataPoint.items.map(item => finalArray.push(item));
    // 		return finalArray;
    // 	}
    // 	finalArray.push(dataPoint);
    // 	return finalArray;
    // }, []);
    // let result = processedDataPoints.length === inputData.length;
    // if (result) {
    // 	processedDataPoints.forEach(dataPoint => {
    // 		console.log("datapoint", dataPoint);
    // 		// Check if every piece of data in the processed data array
    // 		// Contains every key-value pair from the input data provided to the chart
    // 		if (!dataPoint.containsKeysAndValuesFrom(inputData)) {
    // 			result = false;
    // 		}
    // 	});
    // }
    // TODO - Implement Deeper check
    return chartObj.displayData.datasets.length === inputData.datasets.length;
};
// Objects/data
export var selectors = {
    OUTERSVG: "svg.chart-svg",
    INNERWRAP: "g.inner-wrap",
    TOOLTIP: "div.chart-tooltip",
    LEGEND_BTN: "li.legend-btn",
    pie: {
        SLICE: "path"
    }
};
export var colors = [
    "#009BEF",
    "#95D13C",
    "#785EF0",
    "#F87EAC",
    "#FFB000",
    "#00B6CB",
    "#FF5C49",
    "#047CC0",
    "#FE8500",
    "#5A3EC8",
    "#40D5BB",
    "#FF509E"
];
